File: src/main/java/tk/amplifiable/mcgradle/tasks/mc/TaskReobf.java
package club.ampthedev.mcgradle.tasks.mc;

import com.google.common.io.Files;
import net.md_5.specialsource.JarMapping;
import net.md_5.specialsource.JarRemapper;
import net.md_5.specialsource.provider.ClassLoaderProvider;
import net.md_5.specialsource.provider.JarProvider;
import net.md_5.specialsource.provider.JointProvider;
import org.gradle.api.DefaultTask;
import org.gradle.api.file.FileCollection;
import org.gradle.api.tasks.*;
import org.gradle.api.tasks.bundling.Jar;
import club.ampthedev.mcgradle.MCGradleConstants;
import club.ampthedev.mcgradle.Names;
import club.ampthedev.mcgradle.utils.ReobfExceptor;

import java.io.File;
import java.io.IOException;
import java.net.MalformedURLException;
import java.net.URL;
import java.net.URLClassLoader;
import java.nio.charset.Charset;
import java.util.Set;

public class TaskReobf extends DefaultTask {
    @InputFile
    private File input = new File(getProject().getChildProjects().get("generated").getBuildDir(), "/projectCache/recompiled.jar");
    @Input
    private String version = MCGradleConstants.EXTENSION.version;
    @InputFile
    private File preFfJar = new File(MCGradleConstants.CACHE_DIRECTORY, String.format("jars/%s/deobfuscated.jar", version));
    @InputFile
    private File srg = new File(MCGradleConstants.CACHE_DIRECTORY, "jars/" + version + "/mappings/srgs/mcp-notch.srg");
    @InputFile
    private File exc = new File(MCGradleConstants.CACHE_DIRECTORY, "jars/" + version + "/mappings/srgs/mcp.exc");
    @InputFile
    private File methodsCsv = new File(MCGradleConstants.CACHE_DIRECTORY, "jars/" + version + "/mappings/methods.csv");
    @InputFile
    private File fieldsCsv = new File(MCGradleConstants.CACHE_DIRECTORY, "jars/" + version + "/mappings/fields.csv");
    @InputFiles
    private FileCollection libs = getProject().getConfigurations().getByName(Names.MC_DEPENDENCIES_CONF);

    @OutputFile
    private File output = new File(getProject().getChildProjects().get("generated").getBuildDir(), "/projectCache/reobfuscated.jar");

    @TaskAction
    public void doReobfuscate() throws IOException {
        File srg = getSrg();
        {
            ReobfExceptor exceptor = new ReobfExceptor();
            exceptor.toReobfJar = input;
            exceptor.deobfJar = getPreFfJar();
            exceptor.excConfig = getExc();
            exceptor.fieldCSV = getFieldsCsv();
            exceptor.methodCSV = getMethodsCsv();
            File out = new File(getTemporaryDir(), "class_reobf.srg");
            exceptor.doFirstThings();
            exceptor.buildSrg(srg, out);
            srg = out;
        }
        JarMapping mapping = new JarMapping();
        mapping.loadMappings(Files.newReader(srg, Charset.defaultCharset()), null, null,  false);
        JarRemapper remapper = new JarRemapper(null, mapping);
        net.md_5.specialsource.Jar input = net.md_5.specialsource.Jar.init(this.input);
        JointProvider inheritanceProviders = new JointProvider();
        inheritanceProviders.add(new JarProvider(input));
        inheritanceProviders.add(new ClassLoaderProvider(new URLClassLoader(toUrlArray(libs))));
        mapping.setFallbackInheritanceProvider(inheritanceProviders);
        if (!output.getParentFile().exists()) {
            output.getParentFile().mkdirs();
        }
        remapper.remapJar(input, output);
    }

    private URL[] toUrlArray(FileCollection collection) throws MalformedURLException {
        Set<File> files = collection.getAsFileTree().getFiles();
        URL[] array = new URL[files.size()];
        int i = 0;
        for (File f : files) {
            array[i++] = f.toURI().toURL();
        }
        return array;
    }

    @Input
    public String getVersion() {
        return version;
    }

    @InputFiles
    public FileCollection getLibs() {
        return libs;
    }

    @InputFile
    public File getInput() {
        return input;
    }

    @OutputFile
    public File getOutput() {
        return output;
    }

    @InputFile
    public File getPreFfJar() {
        return preFfJar;
    }

    @InputFile
    public File getSrg() {
        return srg;
    }

    @InputFile
    public File getExc() {
        return exc;
    }

    @InputFile
    public File getMethodsCsv() {
        return methodsCsv;
    }

    @InputFile
    public File getFieldsCsv() {
        return fieldsCsv;
    }

    public void setInput(File input) {
        this.input = input;
    }

    public void setVersion(String version) {
        this.version = version;
    }

    public void setPreFfJar(File preFfJar) {
        this.preFfJar = preFfJar;
    }

    public void setSrg(File srg) {
        this.srg = srg;
    }

    public void setExc(File exc) {
        this.exc = exc;
    }

    public void setMethodsCsv(File methodsCsv) {
        this.methodsCsv = methodsCsv;
    }

    public void setFieldsCsv(File fieldsCsv) {
        this.fieldsCsv = fieldsCsv;
    }

    public void setLibs(FileCollection libs) {
        this.libs = libs;
    }

    public void setOutput(File output) {
        this.output = output;
    }
}

File: src/base/kotlin/club/ampthedev/mcgradle/base/tasks/impl/TaskReobf.kt
package club.ampthedev.mcgradle.base.tasks.impl

import club.ampthedev.mcgradle.base.tasks.BaseTask
import club.ampthedev.mcgradle.base.utils.*
import net.md_5.specialsource.Jar
import net.md_5.specialsource.JarMapping
import net.md_5.specialsource.JarRemapper
import net.md_5.specialsource.provider.ClassLoaderProvider
import net.md_5.specialsource.provider.JarProvider
import net.md_5.specialsource.provider.JointProvider
import org.gradle.api.file.FileCollection
import org.gradle.api.tasks.*
import java.io.File
import java.net.URL
import java.net.URLClassLoader

open class TaskReobf : BaseTask(GENERATE_MAPPINGS) {
    @InputFile
    lateinit var input: File

    @InputFile
    lateinit var deobfJar: File

    @InputFile
    lateinit var srg: File

    @InputFile
    lateinit var methodsCsv: File

    @InputFile
    lateinit var fieldsCsv: File

    @InputFiles
    lateinit var libs: FileCollection

    @OutputFile
    lateinit var output: File

    override fun setup() {
        if (!::deobfJar.isInitialized) deobfJar = project.mcgFile(DEOBFED_JAR)
        if (!::srg.isInitialized) srg = project.mcgFile(MCP_NOTCH)
        if (!::methodsCsv.isInitialized) methodsCsv = project.mcgFile(METHODS_CSV)
        if (!::fieldsCsv.isInitialized) fieldsCsv = project.mcgFile(FIELDS_CSV)
        if (!::libs.isInitialized) libs = project.configurations.getByName(CONFIGURATION_MC_DEPS)
    }

    @TaskAction
    fun reobfuscate() {
        /*val exceptor = ReobfExceptor()
        exceptor.toReobfJar = input
        exceptor.deobfJar = deobfJar
        exceptor.fieldCSV = fieldsCsv
        exceptor.methodCSV = methodsCsv
        val out = File(temporaryDir, "class_reobf.srg")
        exceptor.doFirstThings()
        exceptor.buildSrg(srg, out)*/
        val mapping = JarMapping()
        mapping.loadMappings(srg.bufferedReader(), null, null, false)
        val remapper = JarRemapper(null, mapping)
        val input = Jar.init(input)
        val inheritanceProviders = JointProvider()
        inheritanceProviders.add(JarProvider(input))
        inheritanceProviders.add(ClassLoaderProvider(URLClassLoader(libs.toUrlArray())))
        mapping.setFallbackInheritanceProvider(inheritanceProviders)
        prepareDirectory(output.parentFile)
        remapper.remapJar(input, output)
    }

    private fun FileCollection.toUrlArray(): Array<URL> {
        val files = asFileTree.files
        val list = arrayListOf<URL>()
        for (file in files) {
            list.add(file.toURI().toURL())
        }
        return list.toTypedArray()
    }
}